Draft terminology related to API behavior definitions

pabigot opened the pull request:
This PR contains an updated version of proposed terminology related to #18970, hopefully making it easier to comment though review comments.

The formatted document can be read [here](https://github.com/pabigot/zephyr/blob/api/terms/doc/api18970.md).

The content is in markdown because that's most convenient for me.  Assuming we reach consensus we can figure out how it should be integrated into the standard documentation.

pabigot committed:
WIP: doc: definitions for precisely defining API behavior

This is an evolving document used to come to consensus on the set of
defined terms necessary to precisely document the expected behavior of
API calls in various contexts.

It's in Markdown rather than reStructured Text because that's most
convenient for the editor.

This commit archives the original proposal in issue #18970.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: modifications per discussion 2019-11-19

See mbolivar comment at
https://github.com/zephyrproject-rtos/zephyr/issues/18970#issuecomment-555729669
and subsequent discussion.

Add Background section to reference and summarize key concepts from
existing documentation, particularly for thread state and priority and
execution context.

Replaced "blocking" with "yielding" for precision and consistency with
Zephyr thread terminology.

Add Commentary subsections that give a bit of explanation for why
something is important or how concepts relate to each other.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,175 @@
+
+[glossary]: https://docs.zephyrproject.org/latest/glossary.html
+
+### Thread Terminology
```
I'm guessing this is just background context that isn't going into the edits to the glossary in the final PR, right?

In other words, I expect the ultimate outcome of this work to be a set of new entries in the glossary, which cross-reference appropriately into the existing kernel docs which define these terms.

pabigot replied:
This is simply the first step towards the complete requirements of #18970: not just the terminology, but the examples and policies that use that terminology.  Relates to your question about "queued" below: we'll need that term when we discuss the current/desired semantics of an `_async` suffix, perhaps to justify a `_queued` suffix for an API that is asynchronous and non-yielding.

mbolivar replied:
Marking this resolved for now, we can discuss this point further once it's a real docs PR.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,175 @@
+For API behavior we are interested in whether a particular function
+**may**, **must**, or **must not** be invoked from a specific context.
+
+## Definitions
```
In particular, that the terms in this section will be added to the glossary

pabigot replied:
Perhaps, perhaps not.  Note that the thread terminology is an an area of documentation about threads; most of this probably belongs in an area of documentation about API specification and behavior.  To be mooted once we have the terminology down.

mbolivar replied:
Ditto here, will mark resolved as out of scope for this draft PR, final resolution TBD

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,175 @@
+
+### yielding (function)
+
+A function is yielding when it invokes (directly or indirectly) an
```
```suggestion
A function is yielding when it may invoke (directly or indirectly) an
```

The "may" addition is important to cover cases like the hypothetical `clock_control_request()`, which might yield until the clock is ready, and might not yield if the clock is already ready.

pabigot replied:
when speaking of a function generically I meant all paths it might take, so "may" was implicit but your wording is better.

pabigot replied:
Reworded (significantly).

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,175 @@
+
+#### Commentary
+
+Most public API will satisfy this condition; some private API may not.
```
What condition? Thread safety? I would assume most public API is not thread safe. Interrupt safe is another matter. In general I would expect device access APIs don't automatically lock device data, since that would add unnecessary overhead to single-threaded use. It should generally be the caller's responsibility to ensure exclusive access to such data.

pabigot replied:
I'm not sure I agree, given the existence of pre-emptible threads, and the possibility (per #20919) that cooperative threads may get booted out in the middle of whatever they're doing too.

pabigot replied:
Clarification of the concept of *suspendable* in the latest revision supports your position; I've revised the comment.  If the sanctity of cooperative threads gets broken we'll have a problem.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,175 @@
+will yield if the device is already processing an asynchronous
+operation.
+
+Queued operations are rare(? do not exist?) in Zephyr, and require that
```
No objections to the definition of the term, but I guess it goes without saying that each term should have at least one user before being included in the glossary.

pabigot replied:
Agreed.  Marked it TBD and proposed, expanded on the reasoning behind why this is relevant, and suggested it's probably the wrong thing to do.  I expect to remove it before this gets merged.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,175 @@
+
+## Other Rules
+
+* A function that cannot fulfill its contract when invoked from
```
Are you sure the overhead is worth it for this one? I'm not convinced it is -- if we document that it's not OK to call something from ISR context, the results of doing so should be UB, I think.

pabigot replied:
Obviously this needs to be clarified.  I mean that if the function is documented to be callable from an ISR (as, for example, `gpio_pin_write()` should be), then when that generic interface is invoked from an ISR on a particular implementation (an I2C-based GPIO peripheral) that doesn't allow it, the behavior of the API must not only be defined, but be exactly this behavior.  I.e. it is callable, but it may not succeed.

A function that is documented to not be OK to call from an ISR can produce undefined behavior if that restriction is ignored.

pabigot replied:
This has been addressed by introduction of and commentary on *no-wait*.

pabigot committed:
WIP: api18970: refinements to thread and function/operation terminology

Added more reference to existing Zephyr documentation on thread
terminology and behavior, and updated definitions to be more
consistent and refined.

Replaced "yield" with "suspendable" when speaking of functions, and
gave it a more precise definition based on cooperative thread
behavior.  Added "no-wait" as a way to control suspension behavior.
Address "thread-safe" using these concepts.

Defined operation as distinct from function to allow precise
characterization of asynchronous functions.  Comment on why we should
not use "blocking" as a precise term.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Updated with the following edits.  Some previous comments have been marked "resolved".

Added more reference to existing Zephyr documentation on thread terminology and behavior, and updated definitions to be more consistent and refined.

Replaced "yield" with "suspendable" when speaking of functions, and gave it a more precise definition based on cooperative thread behavior.  Added "no-wait" as a way to control suspension behavior. Address "thread-safe" using these concepts.

Defined operation as distinct from function to allow precise characterization of asynchronous functions.  Comment on why we should not use "blocking" as a precise term.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,288 @@
+
+## Definitions
+
+### suspendable (function)
```
We're having trouble finding a word for this that we both like :).

My concern with this latest attempt is that a function which causes the thread to enter the waiting state is suspendable by this definition. This could be confusing, since the reader might thing "suspendable" means "may enter the suspended state".

What about a straightforward "may switch context"?

pabigot replied:
See new update that uses *rescheduling* and refined terms.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,288 @@
+initiate its operation should fail, in many cases by returning the value
+`-EWOULDBLOCK`.
+
+### isr-callable (function)
```
I wonder if you want to generalize this notion to encompass being callable from what the Linux kernel calls "atomic context".

Focusing on ISRs might be a bit short-sighted -- what if we're not in an ISR, but do hold a spinlock in thread context on SMP? Would we need "spinlock-callable" too?

pabigot replied:
That was the thought behind the original question about whether we needed a `k_is_atomic()`.  For now we have only `k_is_in_isr()` and I want two things: (1), a clear definition of what something must promise in order to be unconditionally callable from an ISR (in the latest version that's *reschedule-safe*), and (2) a way to denote that a function allows a parameter to ensure it's callable even when its behavior would normally not be safe (that'd be *no-wait*).

mbolivar replied:
> That was the thought behind the original question about whether we needed a `k_is_atomic()`.

I don't follow that logic, sorry.

The relevant point in the LWN article I keep linking to is that functions can be safe (or not) to call from atomic context, entirely independently of whether or not there exists a predicate that returns true if and only if the current code is running in atomic context.

You seem to be talking about the latter (existence of a predicate). I am talking about the former (whether a code is OK to call from atomic context as a generalization of interrupt context).

What am I missing?

pabigot replied:
The short version is that I'm focusing on terms that I think I understand.  I don't know what to do about atomic.

Can you explain what you mean by "atomic context"?  If the cases we're trying to address are:
* can be safely invoked from an interrupt handler; versus
* can be safely invoked when a spinlock is held

is there really a relationship between them?

Maybe the thing to do is to start with defining the conditions for something being "atomic-callable".  If you agree, please propose something: I have no idea what that would mean that isn't covered by "reschedule-safe".

mbolivar replied:
Resolved offline, I was seeing things that weren't in the document. Sorry for noise.

pabigot committed:
WIP: api18970: more refinements to thread/scheduler terms and behavior

Document what it means for a thread to become unready.

Introduce the concept of a reschedule point as that term is used
within the kernel.  Note existence of meta-irq threads and that ISRs
include a reschedule point.  Replace "suspendable" by a finer-grained
characterization of "rescheduling", accompanied by "reschedule-safe"
and "context-switching" that promise specific behavior of reschedule
points in the common case of cooperative threads.  Define "no-wait"
and "isr-callable" in terms of reschedule concepts.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: minor clarifications

Grammar and style edits, and remove outdated concern about intentional
breakage of the cooperative concept.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,334 @@
+reschedule point (potentially, but not necessarily, context switching to
+a non-current thread).
+
+Note that whether a reschedule-point function will cause a context
```
```suggestion
Note that whether a rescheduling function will cause a context
```

pabigot replied:
done

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,334 @@
+thread](https://docs.zephyrproject.org/latest/reference/kernel/scheduling/index.html#scheduling):
+* In **cooperative** scheduling a thread transitions from *Running* only
+  when it explicitly invokes an operation that makes it unready
+  (i.e. *Waiting* or *Suspended*) or invokes `k_yield()`.
```
```suggestion
  (i.e. *Suspended*, Waiting*, or *Terminated*), or invokes `k_yield()`.
```

mbolivar replied:
Also, actually, is this true?

What if I call `k_thread_abort(some_coop_thread_id)` from another CPU on SMP?

Not super critical to the terminology below, but it seems like that ought to make the aborted thread unready by these definitions.

pabigot replied:
Reworded to acknowledge external changes to thread state.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,334 @@
+are disabled any function may be context-switched if the interrupt
+causes a meta-irq thread to become ready.
+
+### reschedule-safe (function)
```
Did you consider calling this "cooperative-safe", and if so, why did you decide against it?

pabigot replied:
I did not consider that, but it seems to work especially when paired with cooperative-unsafe.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,334 @@
+
+A rescheduling function that is not reschedule-safe is no-wait when
+calls to it can be forced by a parameter into a reschedule-safe path,
+i.e. one that will not trigger a context switch from a cooperative
```
For example, using "cooperative-safe" would probably make this clause unnecessary.

The above would read "a rescheduling function that is not cooperative-safe is no-wait when calls to it can be forced by a parameter into a cooperative-safe path," which I think is pretty clear on its own.

pabigot replied:
Revised to use `cooperative-unsafe`.

mbolivar reviewed (commented):
I'm on board with the synchronous, asynchronous, and context-switching definitions.

The "rescheduling" term reads a bit strangely, but it makes sense and I think we should go with its definition. Choosing an odd word actually probably helps avoid people using the term loosely.

I feel the same way about "no-wait" as I do "rescheduling".

I would like to reach agreement on the isr-callable issue.

pabigot committed:
WIP: api18970: updates to unready, reschedule-safe

Acknowledge that a thread may be made unready externally.  Use "made
unready" as a technical term where appropriate.

Define "context switch" in terms of the behavior at a reschedule
point.

Tighten up the definition of rescheduling.  Replace "reschedule-safe"
with "cooperative-safe" and its inverse "cooperative-unsafe".  Define
no-wait in terms of controllable behavior of a cooperative-unsafe
function.

Update definition of "isr-callable".

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Summary of recent updates:

Document what it means for a thread to become unready.  Acknowledge that a thread may be made unready externally.  Use "made unready" as a technical term where appropriate.

Introduce the concept of a reschedule point as that term is used within the kernel.  Note existence of meta-irq threads and that ISRs include a reschedule point.  Define "context switch" in terms of the behavior at a reschedule point.  Replace "suspendable" by a finer-grained characterization of "rescheduling", accompanied by "cooperative-safe" (vs unsafe) and "context-switching" that promise specific behavior of reschedule points in the common case of cooperative threads.  Define "no-wait" and "isr-callable" in terms of reschedule concepts.

pabigot committed:
WIP: api18970: tweak cooperative-safe

"make unready" doesn't cover k_yield(), so use "transition from
Running" when trying to characterize the behavior that could cause a
cooperative thread to context switch.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: define contexts and contextual restrictions

Based on the answer to issue 21341 identify functions that cannot be
invoked from interrupt (or pre-kernel) context as those that could
cause the invoking thread to "sleep".  Define "sleep" as a voluntary
transition to a non-running active thread state.  Discard the entire
concept of "cooperative-safe" as it is fragile and not actually
relevant to this case.

Generalize the concept of no-wait from preventing a transition to
Waiting to preventing an attempt to sleep.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: move background below definitions

Address the concerns that too much of this is repeating or refining
material that belongs elsewhere by moving the supporting material
after the definitions.  Accept that nobody can understand the
definitions without understanding the supporting material.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Summary of recent updates:

Based on the answer to issue 21341 identify functions that cannot be invoked from interrupt (or pre-kernel) context as those that could cause the invoking thread to "sleep".  Define "sleep" as a voluntary transition to a non-running active thread state.  Discard the entire concept of "cooperative-safe" as it is fragile and not actually relevant.

Generalize the concept of no-wait from preventing a transition to Waiting to preventing an attempt to sleep.

Functions can be called from an ISR if they do not cause their invoking thread to sleep.

Functions are no-wait if they can be prevented from causing their invoking thread to sleep.

pabigot committed:
WIP: api18970: restore missing modifier

no-wait means not blocking.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,408 @@
+to sleep, because without in invoking thread it cannot implement the
+required behavior.
+
+Kernel API assessed using terminology below, plus "cooperative-safe"
```
This table was very helpful. I would not call it a digression or random note, but maybe promote it up higher.

pabigot committed:
WIP: api18970: miscellaneous cleanups

Typos, improved wording.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: refactor to isolate proposed attributes

This effort has identified three attributes that together define
callability and behavior related to thread and interrupt context:
reschedule, sleeps, and no-wait.

A fourth attribute specifies whether the function is decoupled from
the operation it initiates: async

A fifth attribute is identified from work on PR 21389 relating to user
mode programs: supervisor

Name these attributes, and try to describe them in simple standalone
terms in the absence of the full context required to understand why
they're the ones we need.  Some of that context will come from
refactoring and submitting a generalized async notification feature
along with resource management services prototyped in 21090.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: add file listing Zephyr public functions and attributes

Extracted from PR 21389 is the existing documentation on behavior
attributes, which is limited to identifying calling context (Threads
and ISRs) and privileges (Supervisor Mode).  Functions declared in
kernel.h that do not have this information have been added.

Whether the function is a syscall or a macro is indicated by Type.

The Threads, ISRs, and Supervisor (Mode) are from 21389.

A few functions have been extended with the attributes under
consideration.  Presumably anything that is Threads but not ISRs is
sleep in the terminology here.  reschedule is not identified by 21389,
nor is no-wait except when certain functions are marked as callable
from ISRs.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: more revisions

Remove async as an API attribute.  It will be obvious from the
parameters that a function is async.

Remove the extension of no-wait to affect behavior of async.  This was
yet another result of considering "blocking" to imply "sleeps".  There
is no implicit problem with invoking async functions from interrupts,
as long as initialization of the operation cannot sleep.

Add isr-ok which is necessary for generic API that may be called from
interrupts but that invokes sleep functions internally at points where
a no-wait path may not be selected.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,511 @@
+### isr-ok
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context to the function can force an execution
```
Parse error; can you please reword?

pabigot replied:
Something like "can be called from interrupt context ~to the function can~ *and the function will* force an execution".  Basically it's a callee-implemented no-wait.  See details below.

mbolivar replied:
OK... then I think maybe a more complete definition is:

```suggestion
called from interrupt context and function will force an execution
path that will not cause the invoking thread to sleep, regardless
of the values of any of its parameters.
```

The distinction between isr-ok and no-wait seems to be that an isr-ok function is definitionally never going to sleep if called by an ISR (though it might error out), while no-wait can be asked politely not to do so.

Did I understand you?

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,511 @@
+attempt to cause the invoking thread to sleep.
+
+Functions that are **isr-ok** may be invoked from interrupt context, and
+will return an error if they would have tried to sleep.
```
Are you sure about making this a hard universal rule? Why not allow certain functions to conform to the historical Zephyr style of not checking for this type of error except via asserts, to avoid overhead?

pabigot replied:
Because that historical style is exactly why we need this sort of thing.

Take `k_sem_take()`.  You can't call it from an interrupt unless you invoke it in no-wait mode.  If you do the kernel will panic--either because of an assert, or because it's wandered into the fog.  It's "OK" to not return an error because the developer will always know whether the call is in the body of an interrupt handler or a function invoked from a thread, and will have rigorously tested all execution paths with assertions enabled and will be justified in their confidence that the application behaves correctly.

Now consider `gpio_pin_write()`.  You should *always* be able to call that from thread and interrupt context, because it's safe to do that with most SOC controllers and making it illegal would be a major pain in driver code where you need to do something like deassert a signal.

But if the underlying controller identified in a sensor's devicetree node is accessed with I2C then when the driver deasserts the signal from an interrupt handler the GPIO implementation absolutely cannot try to take the semaphore that guarantees mutex of the bus.  So `gpio_pin_write()` must be `isr-ok`, and rather than bring the system down must instead return an error to indicate that it didn't do what it was asked to do.  And the code must propagate that error back to its caller, which must check it and complete its responsibilities some other way.

Now this does reveal an ambiguity that's part of why `k_sem_take()` is not `isr-ok`: because if it was, it would have to do a runtime check, which would get people upset (though the coding guidelines are going to require it anyway).  `gpio_pin_write()` is also not `sleep` or even `reschedule`, even though in certain controllers it will have to sleep.

That's why, once all this stuff gets in place and topic-gpio starts moving forward again, the GPIO API will have to be extended so each gpio controller can provide information about which of its operations are `sleep` so applications can take that into account when interacting with the devices.   (See #17827, which is the fundamental need I've had for the last six months that's motivating this whole effort, and #21090 where the transition operations must be marked as to whether they might sleep so the generic interface can safely avoid panics.)

pabigot commented on doc/api18970.md:
```diff
@@ -0,0 +1,511 @@
+
+name              | reschedule | sleep | no-wait | isr-ok | supervisor
+----------------- | ---------- | ----- | ------- | ------ | ----------
+`k_busy_wait`     | no         | -     | -       | -      | no
```
`-` means it's not specified.  If something is not `reschedule`, it's not `sleep`.  If it's not `sleep`, it's not `no-wait`.  That's a consequence of the way the terms are defined. Anything that isn't `reschedule` is assumed to be callable from thread or interrupt context, with caveats.

At this stage I'm focused on the minimum number of attributes necessary to describe behavior in the case where a specific attribute is present.  We can't assume that if `driver_api_foo()` is *not* marked `reschedule` that it can't be pre-empted due to making another thread ready: maybe the way it has to be implemented requires that freedom.  Absence of attribute is not absence of behavior associated with an attribute.

Also see below on why it's not `isr-ok`.

mbolivar replied:
> If something is not `reschedule`, it's not `sleep`. If it's not `sleep`, it's not `no-wait`.

Gotcha. Isn't the proper value `no` then for `sleep` and `no-wait`, rather than `-`? From your comment, these *are* specified: the short answer is "no", the long answer is "no by definition, given the value of the `reschedule` column".

> Also see below on why it's not isr-ok

Same here. The answer seems like it's "no" rather than "-".

I would personally have found this clearer with "no" used, at least.

pabigot replied:
I feel the distinction between "attribute is not supported because of the way the function is implemented" (and therefore the behavior could be changed to add the attribute) and "attribute is not supported because it's not relevant" is important.  People who don't understand this will have an incomplete understanding of the relationship between the attributes.

So I've marked these `no*` with `*` documenting that point.

I've also updated `isr-ok` and added a couple more attributes.  Please mark resolved any comments you've made that are no longer relevant.

mbolivar replied:
> So I've marked these `no*` with `*` documenting that point.
> 
> I've also updated `isr-ok` and added a couple more attributes.

:+1: 

>  Please mark resolved any comments you've made that are no longer relevant.

Done. The "resolve conversation" button is not available on some of my outdated comments. Please disregard all of those as of time of writing.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,511 @@
+  sleep
+* **no-wait** if a parameter to the function can prevent the invoking
+  thread from trying to sleep
+* **isr-ok** if the function can be safely called from an ISR even if it
```
```suggestion
* **isr-ok** if the function can always be safely called from an ISR even if it
```

mbolivar commented:
@pabigot I think I understand you on all the definitions above. I just have some wording questions to check that and then I think these terms are good and would be useful to have in the official docs.

pabigot committed:
WIP: api18970: wording tweaks

Fill in "no*" for empty cells in the example table, indicating that
the absence is a consequence of the value of another attribute rather
than an implementation decision.

Add pre-kernel-ok as another attribute that extends isr-ok to
explicitly support safe invocation from pre-kernel initialization
stages.

Add async as another attribute in preparation for precise description
of asynchronous APIs.

Clarify distinction between isr-ok and no-wait, and relation between
no-wait and async.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: update table and implicit attribute relations

Make more things "yes*" as a consequent of other attribute values.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,570 @@
+
+#### Commentary
+
+This attribute is of relevance specifically when considering application
```
```suggestion
This attribute is of relevance specifically when considering applications
```

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,570 @@
+### isr-ok
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context.  The function will use `k_is_in_isr()` to
```
```suggestion
called from interrupt context.  If necessary, the function will use `k_is_in_isr()` to
```

It's not necessary for `k_busy_wait()` to do this, for example.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,570 @@
+detect its calling context and force an execution path that will not
+cause the invoking thread to sleep.
+
+This attribute is intended for **sleep** functions that may be
```
```suggestion
This attribute is useful for **sleep** functions that may be
```

Just a suggestion, since again `k_busy_wait()` is isr-ok but not sleep.

In general I think it's a bit cleaner if the definition of the term `isr-ok` doesn't assume that functions so marked are `sleep`, even if that's the primary use case.

pabigot replied:
I've taken the previous suggestion to add "If necessary" regarding `k_is_in_isr()`, but for now I'm going to stick with "intended" because that is its purpose.  "intended" shouldn't imply "reserved for use by".

At this time the only guidance I have for what makes something callable from an interrupt is that "it doesn't sleep".  Nobody's explained what "sleep" means, and my attempts to define it rigorously have led me deep into rabbit warrens so I'm just gonna hand-wave it.

People could tag non-sleep functions with `isr-ok`, but I'm very concerned about a slippery slope where people say "Hey, `k_cycle_get()` (or `k_ms_to_ticks_floor32()` or `crc16()`) isn't marked `isr-ok` so I can't use it in an interrupt?" and we end up explicitly marking a huge number of functions that have nothing to do with thread sleep behavior as `isr-ok`.

If we want complete explicit attribute sets for every API function I'd rather see some tooling that synthesizes derived attributes based on the presence/absence of indicator attributes and populates those in the generated documentation.  Anything that isn't marked `sleeps` should be assumed to be callable from an interrupt.  (Whether doing so is a good idea is a different issue.)

mbolivar replied:
> I'm going to stick with "intended"

ACK

> I'm very concerned about a slippery slope where [...] we end up explicitly marking a huge number of functions that have nothing to do with thread sleep behavior as `isr-ok`.
>
> If we want complete explicit attribute sets for every API function I'd rather see some tooling that synthesizes derived attributes based on the presence/absence of indicator attributes and populates those in the generated documentation. 

OK. Perhaps I'll understand this concern better when you have time to show how you plan on using these terms in a non-draft context to mark up existing APIs in RST documentation.

mbolivar commented on doc/api18970.md:
```diff
@@ -0,0 +1,570 @@
+### pre-kernel-ok
+
+The pre-kernel-ok attribute is used on a function that is also
+**isr-ok** to indicate that it will take reasonable steps to ensure it
```
Looks like the term pre-kernel-ok is a new addition since the last time I looked at this.

Are there any situations in which a function might be pre-kernel-ok, but not isr-ok? I get that in practice they're similar (both "atomic context OK"), but I wonder if it would be better to leave a little wiggle room in the definition by not assuming that `pre-kernel-ok` implies `isr-ok`.

pabigot replied:
Yes; it's there specifically because there's API in work (#21090) that specifically needs to be safe to invoke before the kernel gets started, since one of its use cases is to bring up the system clock.  However, `k_is_in_isr()` is false before the kernel gets started, so API that wants to promote itself as usable in pre-kernel initialization must do an extra check to avoid invoking API that could sleep.  (Nothing can sleep until the kernel gets past pre-kernel initialization and stops running on the interrupt stack.)

So it's not that something might be pre-kernel-ok but not isr-ok, but rather that something might be isr-ok and not pre-kernel-ok because it doesn't do the full check.

I've added "intended" here to parallel the text in **isr-ok**.

I still have no idea what "atomic context" means, and I don't think we should be using that term without a definition, even informally.

mbolivar replied:
> Yes; it's there specifically because there's API in work (#21090)

Yep, I get why you're adding this.

>> wonder if it would be better to leave a little wiggle room in the definition by not assuming that pre-kernel-ok implies isr-ok.
>
> I've added "intended" here to parallel the text in isr-ok.

I take it from this that you are not interested in this wiggle room. Perhaps this relates to your earlier concern about slippery slopes towards excessive markup, which I think I only half understand.

> I still have no idea what "atomic context" means, and I don't think we should be using that term without a definition, even informally.

I'll propose "any execution context in which it is a programming error to invoke an operation that sleeps" as an informal definition, using the terms in this PR.

pabigot replied:
> I take it from this that you are not interested in this wiggle room.

I understood you to be saying "don't assume that pre-kernel-ok implies isr-ok". I don't see that implication in the text, and in fact it's not true.  The text did suggest pre-kernel-ok would be *used* only for explicitly isr-ok functions because you *cannot* assume that isr-ok implies pre-kernel-ok.  I've now taken that out here and will do so in #21678 so you could have a pre-kernel-ok function that's not isr-ok; hopefully that's more like what you had in mind.

> > I still have no idea what "atomic context" means....
> 
> I'll propose "any execution context in which it is a programming error to invoke an operation that sleeps" as an informal definition, using the terms in this PR.

For "atomic" I like [this definition](https://wiki.osdev.org/Atomic_operation):

> An atomic operation is an operation that will always be executed without any other process being able to read or change state that is read or changed during the operation. It is effectively executed as a single step....

There's nothing in this that prevents an atomic operation from being interrupted or sleeping as long as those interruptions don't affect the state of the operation.

What you propose above is a useful definition that covers both pre-kernel and interrupt contexts in Zephyr, but I don't see what we're gaining by using "atomic" in conjunction with "context".  I do see that the term is [used in Linux](https://lwn.net/Articles/274695/) as you'd pointed out before, but that reference seems to be an argument for not using it to mean "can't sleep".  Maybe it should be "sleep-disallowed context".

mbolivar replied:
> but I don't see what we're gaining by using "atomic" in conjunction with "context". I do see that the term is [used in Linux](https://lwn.net/Articles/274695/) as you'd pointed out before, but that reference seems to be an argument for not using it to mean "can't sleep". Maybe it should be "sleep-disallowed context".

I'm happy to stop borrowing the term "atomic context" from Linux -- which I'm only using informally and am not proposing we include in this list of terms -- if you are unhappy with it. I just use it because it's familiar shorthand from Linux, and Zephyr borrows a lot from there.

I would like to have an informal term we can use for "times when you can't sleep," since I think it's a generally useful notion, as shown by the inclusion of pre-kernel-ok and isr-ok in this list of terms.

With apologies to the Beastie Boys, we can call it "on our way to Brooklyn" mode if you want; I don't care :).

mbolivar replied:
> I understood you to be saying "don't assume that pre-kernel-ok implies isr-ok". I don't see that implication in the text, and in fact it's not true. The text did suggest pre-kernel-ok would be _used_ only for explicitly isr-ok functions because you _cannot_ assume that isr-ok implies pre-kernel-ok. I've now taken that out here and will do so in #21678 so you could have a pre-kernel-ok function that's not isr-ok; hopefully that's more like what you had in mind.

Yep, that's clearer to me now, thanks.

pabigot replied:
Can we just call it "non-thread context" and define it as being the union of pre-kernel and interrupt context, rather than relying on the concept of "can't sleep" for its definition?  Those two plus thread context are, AFAIK, the only contexts available.

Note that in the final version I don't try to specify what "sleep" means.  That's because it becomes something like "an operation sleeps when, if it were invoked in thread context, it would cause the running thread to transition itself to a non-running state" (which could be suspended, waiting, ready-but-not-head-of-ready-queue).  Which is horrific, and probably subtly inaccurate.

mbolivar replied:
> Can we just call it "non-thread context" and define it as being the union of pre-kernel and interrupt context, rather than relying on the concept of "can't sleep" for its definition? Those two plus thread context are, AFAIK, the only contexts available.

I'm fine with using the term "non-thread context" between ourselves, but that's not exactly what I'm referring to when I say "atomic context" -- the latter applies even in thread context whenever a spinlock is held.

pabigot replied:
Ah.  In that case I'd want "spin-locked context" as a more useful primitive concept, and you can define "atomic context" as "non-thread or spin-locked".  Though the way I undersatnd "atomic" corresponds more to "spin-locked" alone: interrupts aren't inherently atomic since they can be nested.

AFAIK, you can't sleep in an ISR because it just won't work to sleep while running on the interrupt stack.  You can sleep while holding a spin lock, but there's a chance you'll deadlock.  (ZLIs and possibly meta-IRQ threads becoming enabled break IRQ locks, so you might be able to do it in some perverse but functional way even without SMP).

mbolivar replied:
> AFAIK, you can't sleep in an ISR because it just won't work to sleep while running on the interrupt stack. You can sleep while holding a spin lock, but there's a chance you'll deadlock.

Right, which is why I used "it is a programming error" in my definition, which I believe applies to both cases.

>  In that case I'd want "spin-locked context" as a more useful primitive concept, and you can define "atomic context" as "non-thread or spin-locked".

As far as I know, that's a fine definition for Zephyr on Arm Cortex-M. :thumbsup:

mbolivar reviewed (approved):
Looks good, thanks!

Looks like `pre-kernel-ok` is new. It makes sense but I wonder if we should relax the definition a little bit, to give future changes some room to breathe.

pabigot committed:
WIP: api18970: isr-ok and pre-kernel-ok tweaks

Also a slight rewording to hint at the fact that meta-IRQ threads are
"cooperative" threads that preempt other threads.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP: api18970: remove implied relation between -ok tags

pre-kernel-ok should not be described in a way that suggests it's
coupled to an explicitly isr-ok function.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mbolivar commented:
I am +1 on these terms as of f2474c6. Thanks!

pabigot commented:
Replaced by  #21678
